project13/ViewController.swift

```diff
@@ -4,16 +4,191 @@
+    }()
+    
     override func viewDidLoad() {
         super.viewDidLoad()
-        // Do any additional setup after loading the view.
+        
+        let newTitle: String = "InstaFilter"
+        title = newTitle
+        
+        navigationItem.rightBarButtonItem = UIBarButtonItem(barButtonSystemItem: .add, target: self, action: #selector(importPicture))
+
+        currentFilter = CIFilter(name: "CITwirlDistortion")
+        
+    }
+    @objc private func importPicture() {
+        startLoadingAnimation()
+        DispatchQueue.main.async { [weak self] in
+            guard let self = self else { return }
+            let picker = UIImagePickerController()
+            picker.allowsEditing = true
+            picker.delegate = self
+            self.present(picker, animated: true) {
+                self.stopLoadingAnimation()
+            }
+        }
+    }
+    
+    func imagePickerController(_ picker: UIImagePickerController, didFinishPickingMediaWithInfo info: [UIImagePickerController.InfoKey : Any]) {
+        
+        guard let image = info[.editedImage] as? UIImage else { return }
+        dismiss(animated: true)
+        currentImage = image
+        guard let currentImage = currentImage else { return }
+        
+        let beginImage = CIImage(image: currentImage)
+        currentFilter.setValue(beginImage, forKey: kCIInputImageKey)
+        applyProcessing()
+        stopLoadingAnimation()
+        
+    }
+    
+    @IBAction func changeFilter(_ sender: UIButton) {
+        let alertController = UIAlertController(title: "Choose filter", message: nil, preferredStyle: .actionSheet)
+        CIFilterName.allCases.forEach { filter in
+                  let action = UIAlertAction(title: filter.stringName, style: .default, handler: { _ in
+                      self.set(filter: filter)
+                  })
+                  alertController.addAction(action)
+              }
+        if let popoverController = alertController.popoverPresentationController {
+            popoverController.sourceView = sender
+            popoverController.sourceRect = sender.bounds
+        }
+        present(alertController, animated: true)
+    }
+    
+    @IBAction func save(_ sender: UIButton) {
+        guard let image = imageView.image else {
+            let alertController = UIAlertController(title: "Error", message: "Please selct an image", preferredStyle: .actionSheet)
+            alertController.addAction(UIAlertAction(title: "Ok", style: .destructive))
+            present(alertController, animated: true)
+            return
+        }
+        UIImageWriteToSavedPhotosAlbum(image, self, #selector(image(_:didFinishSavingWithError: contextInfo:)),nil)
+    }
+    
+    @IBAction func intensityChanger(_ sender: Any) {
+        applyProcessing()
+    }
+    
+    private func applyProcessing() {
+        
+       guard let currentImage = currentImage else { return }
+        guard let currentFilter = currentFilter else { return }
```

Comment:
Here you can just go:

```suggestion
        guard let currentImage = currentImage, let currentFilter = currentFilter else { return }
```

Reply:
Done! =)